### PR TITLE
Speed up testing with tox by splitting off flake8

### PR DESCRIPTION
**Pull request by thomwiggers**
Also enables tox -e flake8

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -1,8 +1,16 @@
...
 [testenv]
 deps = -rrequirements-dev.txt
 commands =
     py.test sasstests.py
+
+[testenv:flake8]
+deps= flake8
+skip_install=True
+commands=
     flake8 .
```
Comment (asottile) [inline tox.ini]: To elaborate, we really want to lint against each of the languages targetted as there are often differences in flake8 output depending on python version

Comment (thomwiggers) [inline tox.ini]: So? `flake8` should complain about python 3-only syntax because this project still is compatible with it. Since you dropped 3.3 I'm not aware of any syntax that's valid in 2.7 and not in 3.4+ (`u''` comes to mind for 3.3).

Comment (asottile) [inline tox.ini]: `unicode` for one

Comment (thomwiggers) [inline tox.ini]: Also this doesn't disable `flake8` in the individual CI runs.

Comment (thomwiggers) [inline tox.ini]: Hmm, `unicode` is a good one. Since the other way around (`flake8` only on `py3x`) also isn't a good solution, one possibility would be doing `{py27,py34}-flake8`. This does complicate things slightly.

I normally deal with py3-only projects, which doesn't come with all this hassle 😞 .

Comment (asottile) [inline tox.ini]: If only we didn't have to support python2 :)

Comment (asottile) [inline tox.ini]: I'm also not sure this actually speeds things up.  In experience, usually *decreasing* the number of tox environments is the best way to speed things up.  `setup.cfg` does need fixing it seems to make flake8 even pass in tox with this patch:

```diff
diff --git a/setup.cfg b/setup.cfg
index e3a857c..5c4fcd4 100644
--- a/setup.cfg
+++ b/setup.cfg
@@ -3,4 +3,4 @@ upload_doc = build_sphinx upload_doc
 release = sdist upload build_sphinx upload_doc
 
 [flake8]
-exclude = build,dist,docs,ez_setup.py,upload_appveyor_builds.py
+exclude = .tox,build,dist,docs,ez_setup.py,upload_appveyor_builds.py
```

After that, `flake8 .` returns nearly instantly

I'll try and get some numbers for this to see if it's even worth it

**Comment by thomwiggers** [comment]
Making a py2/3 version of the flake8 is too annoying to be worth it; it adds more complexity to tox.ini than time is saved (I don't appear to be able to do [testenv:*-flake8]).

Meanwhile adding `.tox` to the `[exclude]` list makes the repeating of the `flake8` less painful, and then the main appealing thing that remains in this PR is being able to do `tox -e flake8`. At that point you're not winning much over `flake8 .`.

I'm closing this one and opening one with the `setup.cfg` change.

File: tox.ini (hunk 1 above)
Comment (asottile) [inline tox.ini]: Some numbers so we can quantify this:

## `.tox/py27/bin/flake8 .`
- `real	0m0.892s`
- `real	0m0.896s`
- `real	0m0.889s`

Let's call this 1 second

## `tox -e py27` (without flake8)
- `real	2m0.758s`
- `real	2m4.935s`
- `real	1m59.932s`

Let's say ~120s

## `tox -e flake8`
- `real	0m5.010s`
- `real	0m5.033s`
- `real	0m5.034s`

So let's say that's ~5s

## Extrapolating a bit

(I removed the pypy3 environment in my testing, it's broken due to setuptools / pip)

Given we have 5 environments (pypy is significantly slower, but we'll assume it's the same speed as py27, the py3 environments are also slightly slower for whatever reason but for simplicity we'll assume they're the same).

**Current time:** 120s * 5 + 1s * 5 => 605s
**New time:** 120s * 5 + 5s => 605s

So basically no savings -- if anything it's now a fraction of a second slower (since I rounded the `tox -e flake8` time down and rounded the `.tox/py27/bin/flake8 .` time up).  This time savings is entirely irrelevant compared to the 10 minutes of building C bits though :)

Comment (thomwiggers) [inline tox.ini]: > This time savings is entirely irrelevant compared to the 10 minutes of building C bits though :)

If you like to run flake8 through tox, that's a significant downside: the `skip_install` made sure that didn't happen. But the complexity that adding all the different `[testenv:]` variants adds doesn't really warrant making people not have to set up a venv with `flake8`.
